src/theory/uf/proof_equality_engine.cpp

```diff
@@ -0,0 +1,762 @@
+      return false;
+    }
+    // buffer the step in the fact proof generator
+    ProofStep ps;
+    ps.d_rule = id;
+    ps.d_children = exp;
+    ps.d_args = args;
+    d_factPg.addStep(lit, ps);
+    // add lazy step to proof
+    d_proof.addLazyStep(lit, &d_factPg, false);
+  }
+  // second, assert it to the equality engine
+  Node reason = mkAnd(exp);
+  return assertFactInternal(atom, polarity, reason);
+}
+
+bool ProofEqEngine::assertFact(Node lit,
+                               PfRule id,
+                               Node exp,
+                               const std::vector<Node>& args)
+{
+  Trace("pfee") << "pfee::assertFact " << lit << " " << id << ", exp = " << exp
+                << ", args = " << args << std::endl;
+  Node atom = lit.getKind() == NOT ? lit[0] : lit;
+  bool polarity = lit.getKind() != NOT;
+  // register the step in the proof
+  if (d_pfEnabled)
+  {
+    if (holds(atom, polarity))
+    {
+      // we do not process this fact if it already holds
+      return false;
+    }
+    // must extract the explanation as a vector
+    std::vector<Node> expv;
+    flattenAnd(exp, expv);
+    // buffer the step in the fact proof generator
+    ProofStep ps;
+    ps.d_rule = id;
+    ps.d_children = expv;
+    ps.d_args = args;
+    d_factPg.addStep(lit, ps);
+    // add lazy step to proof
+    d_proof.addLazyStep(lit, &d_factPg, false);
+  }
+  // second, assert it to the equality engine
+  return assertFactInternal(atom, polarity, exp);
+}
+
+bool ProofEqEngine::assertFact(Node lit, Node exp, ProofStepBuffer& psb)
+{
+  Trace("pfee") << "pfee::assertFact " << lit << ", exp = " << exp
+                << " via buffer with " << psb.getNumSteps() << " steps"
+                << std::endl;
+  Node atom = lit.getKind() == NOT ? lit[0] : lit;
+  bool polarity = lit.getKind() != NOT;
+  if (d_pfEnabled)
+  {
+    if (holds(atom, polarity))
+    {
+      // we do not process this fact if it already holds
+      return false;
+    }
+    // buffer the steps in the fact proof generator
+    const std::vector<std::pair<Node, ProofStep>>& steps = psb.getSteps();
+    for (const std::pair<Node, ProofStep>& step : steps)
+    {
+      d_factPg.addStep(step.first, step.second);
+    }
+    // add lazy step to proof
+    d_proof.addLazyStep(lit, &d_factPg, false);
+  }
+  // second, assert it to the equality engine
+  return assertFactInternal(atom, polarity, exp);
+}
+
+bool ProofEqEngine::assertFact(Node lit, Node exp, ProofGenerator* pg)
+{
+  Trace("pfee") << "pfee::assertFact " << lit << ", exp = " << exp
+                << " via generator" << std::endl;
+  Node atom = lit.getKind() == NOT ? lit[0] : lit;
+  bool polarity = lit.getKind() != NOT;
+  if (d_pfEnabled)
+  {
+    if (holds(atom, polarity))
+    {
+      // we do not process this fact if it already holds
+      return false;
+    }
+    // note the proof generator is responsible for remembering the explanation
```

Comment:
we could be doing an eager checking of closed proofs here afterwards, right? Not in this PR, I mean, but in proof-new.

Reply:
Actually I had a discussion about this with @alex-ozdemir 

It seems that arithmetic proofs in CongruenceManager do not have the property of having closed proofs.  I believe I did have this check here, but it was spurious.

Reply:
How come? :)

Reply:
Not sure :)